apps/prairielearn/src/lib/assets.js

```diff
@@ -1,70 +1,90 @@
+const staticNodeModules = require('../middlewares/staticNodeModules');
+const elementFiles = require('../pages/elementFiles/elementFiles');
+
+let assetsPrefix = null;
 
 /**
- * Computes the hash of the given directory and returns the first 16 characters.
+ * Computes the hash of the given path and returns the first 16 characters.
+ * The path can be a file or a directory.
  *
- * @param {string} dir
+ * @param {string} pathToHash
  * @returns {Promise<string>}
  */
-async function hashDirectory(dir) {
-  const { hash } = await hashElement(dir, { encoding: 'hex' });
+async function hashPath(pathToHash) {
+  const { hash } = await hashElement(pathToHash, { encoding: 'hex' });
   return hash.slice(0, 16);
 }
 
-async function computeAssetsHash() {
-  assetsHash = await hashDirectory(path.join(APP_ROOT_PATH, 'public'));
-}
-
-async function computeElementsHash() {
-  elementsHash = await hashDirectory(path.join(APP_ROOT_PATH, 'elements'));
-}
-
-function getPackageVersion(packageName) {
-  try {
-    return require(`${packageName}/package.json`).version;
-  } catch (e) {
-    if (e.code !== 'ERR_PACKAGE_PATH_NOT_EXPORTED') {
-      throw e;
-    }
-
-    // If we can't directly resolve the package's `package.json` file, we'll
-    // `require.resolve` the package itself, and then do a little manipulation
-    // to get the `package.json` path from there.
-
-    // Get the resolved path to the package entrypoint, which will look something
-    // like `/absolute/path/to/node_modules/package-name/index.js`.
-    const pkgPath = require.resolve(packageName);
-
-    // Strip off everything after the last `/node_modules/`, then append the
-    // package name.
-    const nodeModulesToken = '/node_modules/';
-    const lastNodeModulesIndex = pkgPath.lastIndexOf(nodeModulesToken);
-    const pkgJsonPath = path.resolve(
-      pkgPath.slice(0, lastNodeModulesIndex + nodeModulesToken.length),
-      packageName,
-      'package.json'
-    );
-
-    const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf-8'));
-    return pkgJson.version;
-  }
+async function computeCachebuster() {
+  const hashes = await Promise.all([
+    hashPath(path.join(APP_ROOT_PATH, 'public')),
+    hashPath(path.join(APP_ROOT_PATH, 'elements')),
+    hashPath(path.join(REPOSITORY_ROOT_PATH, 'yarn.lock')),
+  ]);
+  const hash = crypto.createHash('sha256');
+  hashes.forEach((h) => hash.update(h));
+  return hash.digest('hex').slice(0, 16);
 }
 
 /**
  * Computes the hashes of directories from which we serve cacheable assets.
  * Should be run at server startup before any responses are served.
+ *
+ * Also initializes the assets compiler.
  */
 module.exports.init = async () => {
-  await Promise.all([computeAssetsHash(), computeElementsHash()]);
+  const cachebuster = await computeCachebuster();
+  assetsPrefix = config.assetsUseCachebuster
+    ? `${config.assetsPrefix}/${cachebuster}`
+    : config.assetsPrefix;
```

Comment:
See comment at the end of the PR description about this - maybe we don't need `assetsUseCachebuster` to be configurable?

Reply:
Excellent point! I agree that we don't need configurability here.